addons/hr_work_entry_contract/models/hr_contract.py

```diff
@@ -51,37 +51,55 @@ def _get_interval_leave_work_entry_type(self, interval, leaves, bypassing_codes)
         return self.env.ref('hr_work_entry_contract.work_entry_type_leave')
 
     def _get_contract_work_entries_values(self, date_start, date_stop):
+        start_dt = pytz.utc.localize(date_start) if not date_start.tzinfo else date_start
+        end_dt = pytz.utc.localize(date_stop) if not date_stop.tzinfo else date_stop
+
         contract_vals = []
         bypassing_work_entry_type_codes = self._get_bypassing_work_entry_type_codes()
+        # {calendar: employees}
+        employees_by_calendar = defaultdict(lambda: self.env['hr.employee'])
+        for contract in self:
+            employees_by_calendar[contract.resource_calendar_id] |= contract.employee_id
+
+        all_resources = self.employee_id.resource_id
+
+        attendances_by_calendar = {
+            calendar: calendar._attendance_intervals_batch(
+                start_dt,
+                end_dt,
+                resources=employees.resource_id,
+                tz=pytz.timezone(calendar.tz)
+            ) for calendar, employees in employees_by_calendar.items()}
+
+        resource_calendar_leaves = self.env['resource.calendar.leaves'].search([
+            ('time_type', '=', 'leave'),
+            # ('calendar_id', '=', self.id), --> Get all the time offs
+            ('resource_id', 'in', [False] + all_resources.ids),
+            ('date_from', '<=', datetime_to_string(end_dt)),
+            ('date_to', '>=', datetime_to_string(start_dt)),
+            ('company_id', '=', self.env.company.id),
+        ])
+        # {resource: resource_calendar_leaves}
+        leaves_by_resource = defaultdict(lambda: self.env['resource.calendar.leaves'])
+        for leave in resource_calendar_leaves:
+            leaves_by_resource[leave.resource_id.id] |= leave
```

Comment:
`read_group(domain, fields=['id:array_agg'], groupby=['resource_id'])` and a browse below ?